src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php

```diff
@@ -162,120 +140,237 @@ private function compileRoutes(RouteCollection $routes, $supportsRedirections)
+     * @param Route  $route                A Route instance
+     * @param string $name                 The name of the Route
+     * @param bool   $supportsRedirections Whether redirections are supported by the base class
+     * @param bool   $hasTrailingSlash     Whether the path has a trailing slash
      *
      * @return string PHP code
      *
      * @throws \LogicException
      */
-    private function compileRoute(Route $route, $name, $supportsRedirections, $parentPrefix = null)
+    private function compileRoute(Route $route, string $name, bool $supportsRedirections, bool $hasTrailingSlash): string
     {
         $code = '';
         $compiledRoute = $route->compile();
         $conditions = array();
-        $hasTrailingSlash = false;
-        $matches = false;
-        $hostMatches = false;
+        $matches = (bool) $compiledRoute->getPathVariables();
+        $hostMatches = (bool) $compiledRoute->getHostVariables();
         $methods = $route->getMethods();
-
         $supportsTrailingSlash = $supportsRedirections && (!$methods || in_array('HEAD', $methods) || in_array('GET', $methods));
-        $regex = $compiledRoute->getRegex();
-
-        if (!count($compiledRoute->getPathVariables()) && false !== preg_match('#^(.)\^(?P<url>.*?)\$\1#'.('u' === substr($regex, -1) ? 'u' : ''), $regex, $m)) {
-            if ($supportsTrailingSlash && '/' === substr($m['url'], -1)) {
-                $conditions[] = sprintf('%s === $trimmedPathinfo', var_export(rtrim(str_replace('\\', '', $m['url']), '/'), true));
-                $hasTrailingSlash = true;
-            } else {
-                $conditions[] = sprintf('%s === $pathinfo', var_export(str_replace('\\', '', $m['url']), true));
-            }
-        } else {
-            if ($compiledRoute->getStaticPrefix() && $compiledRoute->getStaticPrefix() !== $parentPrefix) {
-                $conditions[] = sprintf('0 === strpos($pathinfo, %s)', var_export($compiledRoute->getStaticPrefix(), true));
-            }
-
-            if ($supportsTrailingSlash && $pos = strpos($regex, '/$')) {
-                $regex = substr($regex, 0, $pos).'/?$'.substr($regex, $pos + 2);
-                $hasTrailingSlash = true;
-            }
-            $conditions[] = sprintf('preg_match(%s, $pathinfo, $matches)', var_export($regex, true));
 
-            $matches = true;
+        if ($hasTrailingSlash && !$supportsTrailingSlash) {
+            $hasTrailingSlash = false;
+            $conditions[] = sprintf('%s === $pathinfo', self::export($route->getPath()));
         }
 
-        if ($compiledRoute->getHostVariables()) {
-            $hostMatches = true;
+        if ($route->getCondition()) {
+            $expression = $this->getExpressionLanguage()->compile($route->getCondition(), array('context', 'request'));
+
+            if (false !== strpos($expression, '$request')) {
+                $conditions[] = '($request = $request ?? $this->request ?: $this->createRequest($pathinfo))';
```

Comment:
It looks like we don't have any of the fixtures hitting this code path. Is it expected ?

Reply:
not sure what you mean: this code path is run by the test suite (just verified.)